### ISSUE TITLE
Uncaught Exception

### ISSUE THREAD

**Issue by theofidry** [psalm/psalm-plugin-phpunit#102]
Whilst running Psalm with the PHPUnit plugin on my project, at some point I caused it to break hard although I'm not sure from what exactly:

Full stack trace:

```
PHP Warning:  assert(): assert(null !== $param->type) failed in /path/to/project/vendor/psalm/plugin-phpunit/src/Hooks/TestCaseHandler.php on line 397
Warning: assert(): assert(null !== $param->type) failed in /path/to/project/vendor/psalm/plugin-phpunit/src/Hooks/TestCaseHandler.php on line 397
░Uncaught Exception: Argument 2 passed to Psalm\PhpUnitPlugin\Hooks\TestCaseHandler::Psalm\PhpUnitPlugin\Hooks\{closure}() must be an instance of Psalm\Type\Union, null given, called in /path/to/project/vendor/psalm/plugin-phpunit/src/Hooks/TestCaseHandler.php on line 415
Stack trace in the forked worker:
#0 /path/to/project/vendor/psalm/plugin-phpunit/src/Hooks/TestCaseHandler.php(415): Psalm\PhpUnitPlugin\Hooks\TestCaseHandler::Psalm\PhpUnitPlugin\Hooks\{closure}(Object(Psalm\Type\Union), NULL, false, 0)
#1 /path/to/project/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/ClassAnalyzer.php(963): Psalm\PhpUnitPlugin\Hooks\TestCaseHandler::afterStatementAnalysis(Object(PhpParser\Node\Stmt\Class_), Object(Psalm\Storage\ClassLikeStorage), Object(Psalm\Internal\Analyzer\ClassAnalyzer), Object(Psalm\Codebase), Array)
#2 /path/to/project/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/FileAnalyzer.php(211): Psalm\Internal\Analyzer\ClassAnalyzer->analyze(Object(Psalm\Context), Object(Psalm\Context))
#3 /path/to/project/vendor/vimeo/psalm/src/Psalm/Internal/Codebase/Analyzer.php(340): Psalm\Internal\Analyzer\FileAnalyzer->analyze(NULL)
#4 /path/to/project/vendor/vimeo/psalm/src/Psalm/Internal/Fork/Pool.php(193): Psalm\Internal\Codebase\Analyzer->Psalm\Internal\Codebase\{closure}(4, '/Users/tfidry/P...')
#5 /path/to/project/vendor/vimeo/psalm/src/Psalm/Internal/Codebase/Analyzer.php(406): Psalm\Internal\Fork\Pool->__construct(Array, Object(Closure), Object(Closure), Object(Closure), Object(Closure))
#6 /path/to/project/vendor/vimeo/psalm/src/Psalm/Internal/Codebase/Analyzer.php(269): Psalm\Internal\Codebase\Analyzer->doAnalysis(Object(Psalm\Internal\Analyzer\ProjectAnalyzer), 3)
#7 /path/to/project/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/ProjectAnalyzer.php(639): Psalm\Internal\Codebase\Analyzer->analyzeFiles(Object(Psalm\Internal\Analyzer\ProjectAnalyzer), 3, false, true)
#8 /path/to/project/vendor/vimeo/psalm/src/psalm.php(676): Psalm\Internal\Analyzer\ProjectAnalyzer->check('/Users/tfidry/P...', true)
#9 /path/to/project/vendor/vimeo/psalm/psalm(2): require_once('/Users/tfidry/P...')
#10 {main} in /path/to/project/vendor/vimeo/psalm/src/Psalm/Internal/Fork/Pool.php:357
Stack trace:
#0 /path/to/project/vendor/vimeo/psalm/src/Psalm/Internal/Fork/Pool.php(389): Psalm\Internal\Fork\Pool->readResultsFromChildren()
#1 /path/to/project/vendor/vimeo/psalm/src/Psalm/Internal/Codebase/Analyzer.php(473): Psalm\Internal\Fork\Pool->wait()
#2 /path/to/project/vendor/vimeo/psalm/src/Psalm/Internal/Codebase/Analyzer.php(269): Psalm\Internal\Codebase\Analyzer->doAnalysis(Object(Psalm\Internal\Analyzer\ProjectAnalyzer), 3)
#3 /path/to/project/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/ProjectAnalyzer.php(639): Psalm\Internal\Codebase\Analyzer->analyzeFiles(Object(Psalm\Internal\Analyzer\ProjectAnalyzer), 3, false, true)
#4 /path/to/project/vendor/vimeo/psalm/src/psalm.php(676): Psalm\Internal\Analyzer\ProjectAnalyzer->check('/Users/tfidry/P...', true)
#5 /path/to/project/vendor/vimeo/psalm/psalm(2): require_once('/Users/tfidry/P...')
#6 {main}
(Psalm 4.3.1@2feba22a005a18bf31d4c7b9bdb9252c73897476 crashed due to an uncaught Throwable)
make: *** [psalm] Error 1
```

PHPUnit 9.4.4 & Psalm 4.3.1

**Comment by theofidry** [comment]
Additional info:

<img width="459" alt="Screen Shot 2020-12-10 at 22 50 34" src="https://user-images.githubusercontent.com/5175937/101834091-29c74800-3b3a-11eb-83ba-62e3990c1f2d.png">

The offending code:

```php
    /**
     * @dataProvider   fooProvider
     *
     * @psalm-suppress MissingParamType
     */
    public function testSomething($data): void
    {
```

(I only changed the names, the types are as is).

Psalm config:

```xml
<?xml version="1.0"?>
<psalm
        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
        xmlns="https://getpsalm.org/schema/config"
        xsi:schemaLocation="https://getpsalm.org/schema/config"
        strictBinaryOperands="true"
        ignoreInternalFunctionFalseReturn="false"
        ignoreInternalFunctionNullReturn="false"
        findUnusedVariablesAndParams="true"
        findUnusedCode="true"
        findUnusedPsalmSuppress="true"
        ensureArrayStringOffsetsExist="true"
        ensureArrayIntOffsetsExist="true"
        cacheDirectory=".psalm">

    <plugins>
        <pluginClass class="Psalm\PhpUnitPlugin\Plugin" />
    </plugins>

    <projectFiles>
        <directory name="src"/>
        <directory name="tests"/>

        <ignoreFiles>
            <directory name="vendor"/>
        </ignoreFiles>
    </projectFiles>

    <issueHandlers>
        <ParamNameMismatch>
            <errorLevel type="suppress">
                <directory name="src"/>
            </errorLevel>
        </ParamNameMismatch>

        <UnusedClass errorLevel="suppress"/>
    </issueHandlers>
</psalm>
```

**Comment by weirdan** [comment]
Does adding a type (and removing the suppression) fix the issue?

**Comment by theofidry** [comment]
Yep

**Comment by weirdan** [comment]
@theofidry can you also provide an example of your data provider?

**Comment by theofidry** [comment]
Managed to get a reproducer:

```
final class DummyTest extends AbstractTestCase
{
    /**
     * @dataProvider dummyProvider
     */
    public function test_something($data): void
    {
        $this->addToAssertionCount(1);
    }
}

abstract class AbstractTestCase extends TestCase
{
    /**
     * @return iterable<string, array{mixed}>
     */
    public static function dummyProvider(): iterable
    {
        yield 'closure' => [new stdClass()];
    }
}
```

**Comment by weirdan** [comment]
Ok, reproduced. The crash doesn't happen when you specify the return type as `iterable<string, array<int,mixed>>`, so you can use that as a workaround for the time being.

**Comment by weirdan** [comment]
Fixed in [0.15.0](https://github.com/psalm/psalm-plugin-phpunit/releases/tag/0.15.0).

**Comment by theofidry** [comment]
Works like a charm, thanks!
